src/io/tests/test_InputNode.cpp

```diff
@@ -0,0 +1,56 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2022 QMCPACK developers
+//
+// File developed by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Lab
+//
+// File created by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Lab
+//////////////////////////////////////////////////////////////////////////////////////
+
+#include <variant>
+#include <string_view>
+#include "catch.hpp"
+#include "InputNode.hpp"
+#include "type_traits/template_types.hpp"
+#include "type_traits/variant_help.hpp"
+
+namespace qmcplusplus
+{
+
+struct InputA : public InputNode
+{
+  InputA() = default;
+  InputA(const std::string_view in_name, double in_dA, double in_dB, double in_dC, const std::string_view in_prop)
+      : name(in_name), dA(in_dA), dB(in_dB), dC(in_dC), property(in_prop)
+  {}
+  std::string name{"inputA"};
+  double dA{1.1};
+  double dB{1.2};
+  double dC{1.3};
+  std::string property{"some prop"};
+};
+
+struct InputB : public InputNode
+{
+  InputB() = default;
+  InputB(const std::string& in_name, short in_sA) : name(in_name), sA(in_sA) {}
+  std::string name{"inputB"};
+  short sA{1};
+};
+
+using ChildInput = std::variant<RefW<InputA>, RefW<InputB>>;
+
+TEST_CASE("InputNode_BasicUse", "[io]")
+{
+  UPtrVector<InputNode> input_ownership;
+  std::vector<ChildInput> inputs;
```

Comment:
I don't feel good to keep both input_ownership and inputs around.
Also don't see `variant` brings simplification.

```
enum class {
   DENISTY,
   ESTB
} EstimatorType;

template<typename ENUM>
InputNode
{
   ENUM type;
};

UPtrVector<InputNode<EstimatorType>> input_ownership;
```

In this way, we know the type with some restrictions also. Only need input_ownership.
Also header file only need to see the enum class instead of all the class definition.

Reply:
Header file for EstimatorManagerInput doesn't see all the definitions the way I wrote it forward declaration is fine.  I can make this just use variant, the memory waste isn't probably worth worrying about.

Reply:
No memory concern. My concern was that keeping tracking of two vector separately is error-prone

Reply:
I think that is legitimate and going back and dropping the separate storage works well.  This was a premature optimization.